### ISSUE TITLE
Why is <li> not a BLOCK_TAG?

### ISSUE THREAD

**Issue by mundschenk-at** [Masterminds/html5-php#91]
According to the spec, <code>li</code> can contain any "flow content", i.e. practically anything. Why is it not categorizes a <code>BLOCK_TAG</code>?

**Comment by technosophos** [comment]
Because as defined in the spec, it is not `display: block`. There's a reference in the source code to this resource: http://dev.w3.org/html5/markup/elements.html

**Comment by mundschenk-at** [comment]
OK, let me rephrase that. Why cling to HTML4.01 categories instead of the spec? (<code>display: block</code> is only "typical default display properties", not part of the spec per se). To quote from the Mozilla doc listed in the source:<blockquote>The distinction of block-level vs. inline elements is used in HTML specifications up  to 4.01. In HTML5, this binary distinction is replaced with a more complex set of content categories. The "block-level" category roughly corresponds to the category of flow content in HTML5, while "inline" corresponds to phrasing content, but there are additional categories.</blockquote>

**Comment by technosophos** [comment]
This is why:
https://github.com/Masterminds/html5-php/blob/d60382ea883f886ffdb6605cb2b65146b50e70c1/src/HTML5/Parser/DOMTreeBuilder.php#L346

Because li can have its end tag omitted, it's not included in the BLOCK_TAG list of elements. See http://www.w3.org/TR/html5/syntax.html.

I'm unclear at this point whether you are experiencing a particular issue, or whether you're primarily interested in arguing semantics. I imagine we could change the name of the offending symbol from BLOCK_TAG to something else (suggestion?), but it is a meaningful set used to correctly parse or render HTML5 content.

**Comment by mundschenk-at** [comment]
I'm using HTML5-PHP in an updated version of the WordPress plugin wp-Typography. I'm using what could be called "tag introspection" to handle block-level elements and was surprised that <code>li</code> didn't have mask to check against. I've since worked around this, but the solution is not very pretty (https://github.com/mundschenk-at/wp-typography/commit/f96811df1ce652cf69ed2e6339bb2a2b547d5941). So maybe we could have masks that offer a more granular introspection for cases such as this?

**Comment by technosophos** [comment]
Based on your code, I am guessing that `li` is not going to be the only problem case. `td` and `dt` both follow the same rules (closing tag can be omitted).

I'm not at all opposed to adding more markers on the `Elements.php` masks if that helps. Feel free to generate a PR.

**Comment by mundschenk-at** [comment]
Yes, I've since replaced the check for <code>li</code> with a regex based workaround (filtering the resulting string for control characters. (The specific bug was that I got the text content from the surrounding <code>ul</code> element because the search for a block-level parent stopped one layer to late.) 

I'll have a look at a new marker system, though. In your opinion, would it be better to go for "can contain _flow content_" and "end tag can be omitted" or for <code>LIST_ELEMENT_TAG</code> on the same level as <code>BLOCK_TAG</code>, i.e. based suggested default <code>display</code> modes?

**Comment by mattfarina** [comment]
If you look at the [`Elements.php`](https://github.com/Masterminds/html5-php/blob/d60382ea883f886ffdb6605cb2b65146b50e70c1/src/HTML5/Elements.php) file you'll see there are masks that can be tested elsewhere. The `li` is only listed as a known element. "I think" `li`, `td`, and `dt` need to be updated to be block elements and a new mask needs to be created to auto-close them (or optional closing). Then this can be used in the logic as needed.

This is just my thinking off the cuff.

**Comment by technosophos** [comment]
My preference, @mundschenk-at, would be to use terminology closer to the standard, and to focus on lexical attributes (rather than display, which we don't really do). So "flow content" and "omit end tag" are probably better concepts to work with. Does that sound right to you?

**Comment by mundschenk-at** [comment]
Yes, it does.
